cpp/src/inflight_requests.cpp

```diff
@@ -21,11 +21,16 @@ void InflightRequests::insert(std::shared_ptr<Request> request)
   _inflightRequests->insert({request.get(), request});
 }
 
-void InflightRequests::merge(InflightRequestsMapPtr inflightRequestsMap)
+void InflightRequests::merge(InflightRequestsMapPtrPair inflightRequestsMapPtrPair)
 {
-  std::lock_guard<std::mutex> lock(_mutex);
-
-  _inflightRequests->merge(*inflightRequestsMap);
+  {
+    std::lock_guard<std::mutex> lock(_mutex);
+    _inflightRequests->merge(*(inflightRequestsMapPtrPair.first));
+  }
+  {
```

Comment:
**question:** What are the preconditions on the two maps to this function?

Is it the case that every to-cancel request must have been already been observed? Is it allowed to have entries in the "new" requests that are also in the "to-cancel" requests?

Reply:
> Is it the case that every to-cancel request must have been already been observed?

Yes, a request will only be moved to the `_trackedRequests->_canceling` if it was previously in `_trackedRequests->_inflight` and `cancelAll()` was called, see [this block](https://github.com/rapidsai/ucxx/pull/153/files#diff-5f8b26cd809dc146d0f147fe15bee78cf7a1d9ac5898f09e4fa733f34710fc51R109-R133).

> Is it allowed to have entries in the "new" requests that are also in the "to-cancel" requests?

No, the only place where `_trackedRequests->_canceling` is populated is the `cancelAll()` block linked above, thus a request cannot go into "canceling" state right away. The exception is when merging requests, the "canceling requests" will still be "canceling requests" when we merge `InflightRequests` from an `Endpoint` into that of the `Worker`, but in that case they were already being canceled so it's not as if they were immediately registered in a "canceling" state when `Endpoint::registerInflightRequest()` was first called.